### ISSUE TITLE
Compatability with Future Versions of Firefox

### ISSUE THREAD

**Issue by tomrittervg** [xiaoxiaoflood/firefox-scripts#64]
Hi; I'm a Firefox developer on the security engineering team. We're working to lock down unexpected behavior in the browser, and have seen evidence that these scripts are causing unexpected JS loads in the parent process.  That's to be expected, that's exactly what they're going to to do - _but_ the loads are happening _despite_ us turning off our checking behavior if we think these scripts might be present.

I'm hoping you can help us figure out how this may be happening so we can avoid breaking your scripts when we start to enforce the checking behavior in the future (which would stop the script from loading, and silently break stuff.)

Specifically we check the `general.config.filename` pref on the main thread. If you're using some sort of userChromeJS or boostrapLoader that doesn't set that pref _or_ you're loading extensions that ultimately use workers and load a JavaScript file off-main-thread we would start blocking things.  

In the last ~1 month I've seen someone (literally one Windows user) triggering this using `BackTrack Tab History` on Nightly, but with users disabling Telemetry and the non-Windows reporting less verbose, it's possible this is happening more often.

So I was wondering:

1) Can you enable legacy stuff without setting `general.config.filename`? If yes, would you consider making setting this pref a requirement for these scripts to operate?
2) Do you know if any of the extensions (especially `BackTrack Tab History`) load scripts off-the-main thread (e.g. in workers)?

**Comment by 117649** [comment]
Do you mean something like `let worker = new window.Worker('{script url}')`

**Comment by tomrittervg** [comment]
> Do you mean something like `let worker = new window.Worker('{script url}')`?

Yes, that would mean you're using a worker in the extension and is probably indicative of tripping our alerts.  If that's a common use in these extensions we (Mozilla) could look at `general.config.filename` on all threads instead of just main thread....

**Comment by 117649** [comment]
https://github.com/117649/Tab-Groups/blob/9c651d7cf72e9139b883f3bfbb19c3ba5eb416be/resource/modules/FavIcons.jsm#L202-L259
One place I know.

Would [load frame script](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Multiprocess_Firefox/Message_Manager/Frame_script_loading_and_lifetime) also tripping your alerts?
That's a way more common use.

**Comment by AlexVallat** [comment]
Firstly, @tomrittervg , I'd just like to say thank you for reaching out with this. Even if Firefox does end up breaking our addons again, the fact that you've taken the time to look into the issue means a lot.

I'm the author of BackTrack Tab History, and can confirm that it does not use web workers, but it does use (delay load) frame scripts, loaded via `globalMessageManager.loadFrameScript`. It also uses messaging through `globalMessageManager.addMessageListener` and `sendAsyncMessage` both from and to the frame script.

**Comment by xiaoxiaoflood** [comment]
Hi,

Thank you for gently coming here to discuss a way to not break our scripts/extensions even though they aren't officially supported by Mozilla. I really appreciate it. :blush:

> 1. Can you enable legacy stuff without setting `general.config.filename`?

Only through [bootstrapLoader.xpi](https://github.com/xiaoxiaoflood/firefox-scripts/tree/master/extensions/bootstrapLoader), a WebExtension Experiment to load bootstrapped legacy addons. But I don't really like it and I don't use it. userChromeJS (through `general.config.filename`) is better as it runs earlier so it has more powers. Also, userChromeJS can be installed on any Fx version (if you set `general.config.sandbox_enabled = false`), while Experiments are restricted to Nightly/DevEd/unbranded builds.

I didn't write a WebExt Experiment to load userChromeJS scripts and I fear some of them may not work properly that way, like [Master Password+](https://github.com/xiaoxiaoflood/firefox-scripts/blob/master/chrome/masterPasswordPlus.uc.js) that needs to lock all Fx windows (not just `browser.xhtml`) on startup before any content is displayed.

> If yes, would you consider making setting this pref a requirement for these scripts to operate?

`general.config.filename` is already required to use userChromeJS (aka [autoconfig](https://mike.kaply.com/2012/03/22/customizing-firefox-advanced-autoconfig-files/), the JS equivalent to userChrome.css, that's why we call it userChromeJS and use `chrome` folder).

userChromeJS not only enables UserScripts (`.uc.js` files), I also added the bootstrapLoader code in it. That's what I use on my end, I don't use `bootstrapLoader.xpi`.

I don't use nor recommend using WebExt Experiment to enable legacy extensions/scripts, so as long as we can continue having userChromeJS with full privileges I'm fine. So no problem for me if you block Experiments when `general.config.filename` is not set.

**Comment by xiaoxiaoflood** [comment]
> 2. Do you know if any of the extensions (especially BackTrack Tab History) load scripts off-the-main thread (e.g. in workers)?

Most of them use `messageManager`, like DownThemAll!. Some userChromeJS scripts too.

**Comment by tomrittervg** [comment]
Presently my dataset is restricted to Nightly, so it may be coming primarily from bootstrapLoader.xpi.  I'm going to propose checking for `xpinstall.signatures.required = false` and if so, disable our checks.  Hopefully that quiets down the activity. It's preferable to get the data down to near-zero, that means that we understood what the problem was and there's less chance of breaking something unknown!

The other possibility is workers; but given the example provided actually uses the chrome:// scheme (and we allow those) I'm feeling less sure of this. Still; I don't think it would be hard to introduce a static variable to check general.config.filename off-main-thread.  (TBH I'm not sure why I didn't do that in the beginning.)

Finally, I'm going to try to increase our Telemetry to Beta and eventually Release. Enabling Telemetry - if you disabled it - will help us to avoid breaking your stuff. :)  The data I'm specifically collecting would be shown in the 'Events' section of about:telemetry with the key security.javascriptLoad (and security.evalUsage but that one has graduated to enforcement mode so if it was going to break something, it already did.)  Feel free to ping me in the future to follow up on this - and if things did break for you, mozregression is a good tool to figure out what it was!

**Comment by xiaoxiaoflood** [comment]
Thanks. I've been using mozregression extensively since Fx 57 to track and fix breakages. Like your [1570681](https://bugzilla.mozilla.org/show_bug.cgi?id=1570681) did in Fx 71 for bootstrapLoader.xpi users. :stuck_out_tongue_winking_eye:

**Comment by xiaoxiaoflood** [comment]
So Fx 88 broke DownThemAll!:
https://github.com/downthemall/downthemall-legacy/blob/84160d789062691de0ffd82bdc4ebc123b540203/modules/manager/chunk.js#L17-L38

I could simply rename `Services.tm.newThread(0)` by `Services.tm.newNamedThread('')`, but... [JS doesn't have access to it](https://searchfox.org/mozilla-central/rev/368607c4cd5be547021945e4ae60e8eb4365b3c4/xpcom/threads/nsIThreadManager.idl#70). 😕

**Comment by onemen** [comment]
@117649,

If you are using Mac then you can only load extensions from folder in `Extensions` folder
```
; Per-user and system-wide Extensions dir
  (allow file-read*
      (home-regex "/Library/Application Support/[^/]+/Extensions/")
      (regex "^/Library/Application Support/[^/]+/Extensions/"))
```
https://github.com/mozilla/gecko-dev/blob/f8576fec48d866c5f988baaf1fa8d2f8cce2a82f/security/sandbox/mac/SandboxPolicyContent.h#L283-L286

You can add a file with your extension ID in the profile extensions folder that point to another folder.

To test TabMixPlus, I added this line to a file with then name `{dc572301-7619-498c-a57d-39143191b318}`
`~/Library/Application Support/Mozilla/Extensions/TabMixPlus/addon`

I drop this file to any profile that I'm testing, and all the files for developing TabMixPlus are in the repository in
 `~/Library/Application Support/Mozilla/Extensions/TabMixPlus`

**Comment by 117649** [comment]
> @117649,
> 
> If you are using Mac then you can only load extensions from folder in `Extensions` folder
> 
> ```
> ; Per-user and system-wide Extensions dir
>   (allow file-read*
>       (home-regex "/Library/Application Support/[^/]+/Extensions/")
>       (regex "^/Library/Application Support/[^/]+/Extensions/"))
> ```
> 
> https://github.com/mozilla/gecko-dev/blob/f8576fec48d866c5f988baaf1fa8d2f8cce2a82f/security/sandbox/mac/SandboxPolicyContent.h#L283-L286
> 
> You can add a file with your extension ID in the profile extensions folder that point to another folder.
> 
> To test TabMixPlus, I added this line to a file with then name `{dc572301-7619-498c-a57d-39143191b318}` `~/Library/Application Support/Mozilla/Extensions/TabMixPlus/addon`
> 
> I drop this file to any profile that I'm testing, and all the files for developing TabMixPlus are in the repository in `~/Library/Application Support/Mozilla/Extensions/TabMixPlus`

Never used a Mac.

**Comment by onemen** [comment]
@xiaoxiaoflood

Mozilla fixed the issue ([Bug 1785278](https://bugzilla.mozilla.org/show_bug.cgi?id=1785278), [fb762c3](https://github.com/canonical/firefox-snap/pull/8)) with Firefox snap for Linux, by adding an new path to add config files.
Starting with **Firefox 108** (currently in Nightly) you can add the files extracted from [this zip file](https://raw.githubusercontent.com/xiaoxiaoflood/firefox-scripts/master/fx-folder.zip) to these locations:
- copy `config.js` to `/etc/firefox/config.js`
- copy `config-prefs.js` to `/etc/firefox/defaults/pref/config-prefs.js`

You may have to create the sub folders `/defaults/pref`
